src/common/threads_manager.h

```diff
@@ -0,0 +1,157 @@
+/*!
+ * Copyright 2022 by Contributors
+ * \file threads_manager.h
+ * \brief Helper class for control threads in partition builder
+ */
+#ifndef XGBOOST_COMMON_THREADS_MANAGER_H_
+#define XGBOOST_COMMON_THREADS_MANAGER_H_
+
+#include <utility>
+#include <vector>
+#include <unordered_map>
+
+namespace xgboost {
+namespace common {
+
+struct Slice {
+  uint32_t* addr {nullptr};
+  uint32_t b {0};
+  uint32_t e {0};
+  uint32_t Size() const {
+    return e - b;
+  }
+};
+
+template <class DataT>
+class AbstractBuffer {
```

Comment:
to escape using virtual methods i suggest using CRTP here, as example

template <class T, class Impl>
class base {
public:    
    int idx() { return static_cast<Impl*>(this)->idx(); }
    int get_next_item() { counter_++; return idx(); }

protected:    
    base(std::vector<T>* vec_ptr) : vec_ptr_(vec_ptr), counter_(0) {}

    std::vector<T>* vec_ptr_;
    int counter_;
};

template <class T>
class sview : public base<T, sview<T>> {
public:    
    sview(std::vector<T>* vec_ptr) : base<T, sview<T>>(vec_ptr) {}

    int idx() { return ...; }
};